app/spec/utils/utils.spec.js

```diff
@@ -129,4 +132,31 @@ describe("Utils", function() {
     // verif: undefined remains undefined
     expect(obj).toEqual(undefined);
   });
+
+  it("should substitute aliases present in an instance definition.", function() {
+    const tests = require(path.resolve("spec/utils/testUtils"));
+
+    // setup
+    const config = tests.config();
+    const db = proxyquire(cst.DBPATH, tests.stubs());
+
+    var instanceDef = db.getInstanceDefinition(
+      "cacb5448-46b0-4808-980d-5521775671c0"
+    );
+    var substitutedInstanceDef = utils.substituteAliases(instanceDef);
+
+    // verif
+    expect(substitutedInstanceDef.deployment.hostDir).toEqual(
+      "/var/docker-volumes/cacb5448-46b0-4808-980d-5521775671c0"
+    );
+
+    instanceDef = db.getInstanceDefinition("nonexisting-instance-uuid");
+    substitutedInstanceDef = utils.substituteAliases(instanceDef);
+
+    // verif
+    expect(substitutedInstanceDef).toEqual(instanceDef);
+
+    // after
+    tests.cleanup();
+  });
```

Comment:
In fact `utils` should not be tested with framework sensitive tests. Those tests are rather intended for higher level scripts.
`utils` is a library of standalone functions that should be testable in isolation.

I would simply test the following atomic cases (but in one test case):
1. Pass it an instance def with an aliased host dir and a UUID - and check host dir is substituted as expected.
1. Demonstrate that the substitution throws an error if the UUID is missing/empty (in lodash terms.)
1. Demonstrate that host dir is unchanged when it's not aliased, actually even if the UUID is missing/empty.

_(30 min later...)_
I was searching a little and found this: https://www.npmjs.com/package/string#--templatevalues-open-close
I think that's our guy ;-)
Having said that, it will make point 3 above very difficult to implement, so let's forget about that. Let's assume that if one attempts to substitute aliases, then the necessary references must be at hand within the instance def, otherwise: error thrown.

Reply:
>Pass it an instance def with an aliased host dir and a UUID - and check host dir is substituted as expected.

More precisely:

>Pass it an instance def

That is the reason why I end up using the framework tests.
Because it seems easier to get testing resources that way. I could bring my own instanceDef resource, but then that would easily be out of sync if the instanceDef changes don't you think?

After all, this function should probably not be a utils function, since it is made to manipulate instanceDef specifically, I guess not generic enough to be in utils. What do you think?

Reply:
You could pass the piece of the instance def. that is strictly necessary to do the validation. So basically stuff like:
```javascript
var instanceDef = {
  "uuid": "dcb395ac-5594-4ca4-b2ba-1ab60e8694b3",
  "deployment" {
    "hostDir": "/var/docker-volumes/{{uuid}}"
  }
}
```
```javascript
var badInstanceDef = {
  "deployment" {
    "hostDir": "/var/docker-volumes/{{uuid}}"
  }
}
```
```javascript
var awkwardInstanceDef = {
  "deployment" {
    "hostDir": "/var/docker-volumes/dcb395ac-5594-4ca4-b2ba-1ab60e8694b3"
  }
}
```

>After all, this function should probably not be a utils function, since it is made to manipulate instanceDef specifically, I guess not generic enough to be in utils. What do you think?

It's true that it should only be used in p3. You could possibly create a **commons.js** in p3 that exports an API that is needed across p3. That would align p3 on p1 with the existence of an ad-hoc 'commons' API but inside each pipeline's code base.